external-crates/move/crates/move-compiler/src/hlir/match_compilation.rs

```diff
@@ -0,0 +1,1903 @@
+            .take(n)
+            .collect()
+    }
+
+    fn find_counterexample_bool(
+        context: &mut Context,
+        matrix: PatternMatrix,
+        arity: u32,
+        ndx: &mut u32,
+    ) -> Option<Vec<CounterExample>> {
+        let literals = matrix.first_lits();
+        assert!(literals.len() <= 2, "ICE match exhaustiveness failure");
+        if literals.len() == 2 {
+            // Saturated
+            for lit in literals {
+                if let Some(counterexample) =
+                    find_counterexample(context, matrix.specialize_literal(&lit).1, arity - 1, ndx)
+                {
+                    let lit_str = format!("{}", lit);
+                    let result = [CounterExample::Literal(lit_str)]
+                        .into_iter()
+                        .chain(counterexample)
+                        .collect();
+                    return Some(result);
+                }
+            }
+            None
+        } else {
+            let (_, default) = matrix.default();
+            if let Some(counterexample) = find_counterexample(context, default, arity - 1, ndx) {
+                if literals.is_empty() {
+                    let result = [CounterExample::Wildcard]
+                        .into_iter()
+                        .chain(counterexample)
+                        .collect();
+                    Some(result)
+                } else {
+                    let mut unused = BTreeSet::from([Value_::Bool(true), Value_::Bool(false)]);
+                    for lit in literals {
+                        unused.remove(&lit.value);
+                    }
+                    let result = [CounterExample::Literal(format!(
+                        "{}",
+                        unused.first().unwrap()
+                    ))]
+                    .into_iter()
+                    .chain(counterexample)
+                    .collect();
+                    Some(result)
+                }
+            } else {
+                None
+            }
+        }
+    }
+
+    fn find_counterexample_builtin(
+        context: &mut Context,
+        matrix: PatternMatrix,
+        arity: u32,
+        ndx: &mut u32,
+    ) -> Option<Vec<CounterExample>> {
+        // For all other non-literals, we don't consider a case where the constructors are
+        // saturated.
+        let literals = matrix.first_lits();
+        let (_, default) = matrix.default();
+        if let Some(counterexample) = find_counterexample(context, default, arity - 1, ndx) {
+            if literals.is_empty() {
+                let result = [CounterExample::Wildcard]
+                    .into_iter()
+                    .chain(counterexample)
+                    .collect();
+                Some(result)
+            } else {
+                let n_id = format!("_{}", ndx);
+                *ndx += 1;
+                let lit_str =
+                    format_oxford_list!("or", "{}", literals.into_iter().collect::<Vec<_>>());
```

Comment:
I worry that this list may get incredibly long and become unreadable quickly

Reply:
This is what Rust does 🤷

Reply:
Maybe truncate it after three elements? (see e.g., https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=480ca4f850eb061cc28a594af31e7853)

Reply:
I did this, but really we should come back in and do ranges at some point like Rust does

Reply:
Yea totally agreed around adding in ranges (but that's for a future date -- I think this is fine for now).